Add support for more dtypes for serialization

bt2513 opened the pull request:
Fixes #131654

Hi, I noticed the issue above and wanted to contribute. I am still getting familiar with the Pytorch repo and thought it could be nice to gain more insights on how dtypes are supported in Pytorch 🙌

@angelayi, you mentioned in the issue that numbers in `schema.py` should match the numbers [here](https://github.com/pytorch/pytorch/blob/cab90b0049f75a1c6d9ca034e5011fa170a944e8/c10/core/ScalarType.h#L56-L93). I wasn't sure to understand because BFloat16 is next to the number 16 in this list (vs 13 in `schema.py`), did you perhaps mean the order in this [list](https://github.com/pytorch/pytorch/blob/cab90b0049f75a1c6d9ca034e5011fa170a944e8/c10/core/ScalarType.h#L120-L137) instead?
Also, please let me know if I didn't increment the `SCHEMA_VERSION` correctly

Thank you!

@angelayi 

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @penguinwu

bt2513 committed:
support float8_e4m3fn serialization

bt2513 committed:
support float8_e4m3fn serialization

bt2513 committed:
fix numbering

bt2513 committed:
fix numbering

bt2513 committed:
fix typo

bt2513 committed:
fix typo

bt2513 commented:
Hi @angelayii, I just wanted to follow up on this in case you missed it. Thanks :)

angelayi commented on torch/_export/serde/schema.py:
```diff
@@ -8,7 +8,7 @@
 
 # NOTE: Please update this value if any modifications are made to the schema
-SCHEMA_VERSION = (7, 1)
+SCHEMA_VERSION = (7, 2)
```
I don't think we need to bump the schema version

angelayi commented on torch/_export/serde/schema.py:
```diff
@@ -27,6 +27,10 @@ class ScalarType(IntEnum):
+    FLOAT8_E5M2 = 14
+    FLOAT8_E4M3FN = 15
+    FLOAT8_E5M2FNUZ = 16
+    FLOAT8_E4M3FNUZ = 17
```
Can we stick to the numbering [here](https://github.com/pytorch/pytorch/blob/cab90b0049f75a1c6d9ca034e5011fa170a944e8/c10/core/ScalarType.h#L56-L93) (23-26). We should fix the other numberings eventually too..

angelayi reviewed (commented):
Thanks for the bump!

bt2513 committed:
address feedback

bt2513 committed:
address feedback

bt2513 commented:
Thanks for your feedback @angelayi I addressed your comments. Should I go ahead and fix the other numberings too? In that case, what number should I use for the [`UNKNOWN`](https://github.com/pytorch/pytorch/blob/cab90b0049f75a1c6d9ca034e5011fa170a944e8/torch/_export/serde/schema.py#L16) ScalarType?

angelayi commented:
@bt2513 I'm hesitant to fix the other numberings due to BC issues, so let's just keep this for now. Thanks!

bt2513 commented:
@angelayi I tried to resolve the CI issues by running the `python scripts/export/update_schema.py` command, as mentioned in the failure message, but ran into the error:
```
  File "/Users/bertrandthia/Desktop/projects/pytorch/scripts/export/update_schema.py", line 47, in <module>
    next_version, reason = schema_check.check(commit, args.force_unsafe)
  File "/Users/bertrandthia/Desktop/projects/pytorch/torch/_export/serde/schema_check.py", line 243, in check
    if "default" not in d and kind == "struct":
TypeError: argument of type 'int' is not iterable
```
After looking deeper into it, it seems to be because the following [line](https://github.com/pytorch/pytorch/blob/50ed6ce2777f30fcc7f3afdffa00fa6a73b84e03/torch/_export/serde/schema_check.py#L241) defines `fields` as `{'FLOAT8_E5M2': 23, 'FLOAT8_E4M3FNUZ': 26, 'FLOAT8_E4M3FN': 24, 'FLOAT8_E5M2FNUZ': 25}`. As a result, the next lines trigger an error because they try to iterate on the integer 23 (i.e. attempt to test the condition `if "default" not in 23`, but 23 is an int not iterable)

Could you please provide some guidance on how we should handle this? Thanks!

bt2513 commented:
Hi @angelayi, just bumping this :) Please let me know your thoughts on this when you have time 🙌

angelayi commented:
@bt2513 you have to run `python scripts/export/update_schema.py --prefix <path_to_torch_development_diretory>`

bt2513 commented:
> @bt2513 you have to run `python scripts/export/update_schema.py --prefix <path_to_torch_development_diretory>`

After running this command, I actually encountered the issue I had previously described @angelayi

bt2513 commented:
Hi @angelayi - I just wanted to follow up and see if you had any guidance about the above? Could it be that I've missed something in my setup, or is there another explanation for the issue otherwise? Thanks 🙏

bt2513 commented:
Hi @angelayi - bumping this in case you missed my previous comment

bt2513 commented:
Hi @angelayi - bumping this :)

malfet commented on torch/_export/serde/schema.py:
```diff
@@ -27,6 +27,10 @@ class ScalarType(IntEnum):
     COMPLEXDOUBLE = 11
     BOOL = 12
     BFLOAT16 = 13
+    FLOAT8_E5M2 = 23
+    FLOAT8_E4M3FN = 24
+    FLOAT8_E5M2FNUZ = 25
+    FLOAT8_E4M3FNUZ = 26
```
Why there is a gap in numbering?

bt2513 replied:
It's because the numbers need to match what's [here](https://github.com/pytorch/pytorch/blob/cab90b0049f75a1c6d9ca034e5011fa170a944e8/c10/core/ScalarType.h#L80-L83)! Thanks for helping @malfet, I was stuck

malfet commented:
@pytorchbot rebase

kit1980 commented:
Are there plans to finish this?
There are only couple of days left to cherry-pick for 2.5, we probably should remove the milestone.

bt2513 commented:
@pytorchbot rebase

bt2513 commented:
Hi @kit1980 -- I can pick this up again and work on resolving the issues, but I would appreciate some guidance if possible 🙏. I followed up several times in August for help, but unfortunately, I didn’t receive a response. I’ll take a fresh look at this in the coming week to see if I can make progress. Otherwise, it might be better to close this PR and have someone more knowledgeable take over I suppose

bt2513 committed:
Update schema.py

bt2513 committed:
Merge branch 'bthia/support-float8_e4m3fn' of https://github.com/bt2513/pytorch into bthia/support-float8_e4m3fn

bt2513 commented:
Still running into the same issue:
```
(pytorch) bertrandthia@MacBook-Pro-7 pytorch % python scripts/export/update_schema.py --prefix=/Users/bertrandthia/Downloads/ai-journey/pytorch/
Traceback (most recent call last):
  File "/Users/bertrandthia/Downloads/ai-journey/pytorch/scripts/export/update_schema.py", line 45, in <module>
    next_version, reason = schema_check.check(commit, args.force_unsafe)
  File "/Users/bertrandthia/Downloads/ai-journey/pytorch/torch/_export/serde/schema_check.py", line 243, in check
    if "default" not in d and kind == "struct":
TypeError: argument of type 'int' is not iterable
```
Not sure how to resolve. Any guidance would be appreciated @angelayi @malfet @kit1980 🙏

As I mentioned previously above, the error seems to occur when the following [line](https://github.com/pytorch/pytorch/blob/50ed6ce2777f30fcc7f3afdffa00fa6a73b84e03/torch/_export/serde/schema_check.py#L241) is executed by the script. At this point, the variable `fields = {'FLOAT8_E5M2': 23, 'FLOAT8_E4M3FNUZ': 26, 'FLOAT8_E4M3FN': 24, 'FLOAT8_E5M2FNUZ': 25}` is assigned, but this is not a valid value for the rest of the code to run. I am unsure what changes I need to make for it to work, and I find it odd that the script runs with the existing examples but fails with mine when I expand the list similarly

kit1980 commented:
I've removed the milestone as this is not ready for 2.5 release.

Hopefully someone knowledgeable in the area will chime in to help to move this.
